### PR TITLE
Wrw odd image

### PR DESCRIPTION
**Pull request by carlushuang**
wrw odd image pad is along gemm_k direction. it is different from fwd/bwd, need careful figure out move_slice length along each direction, after such pad.

### CODE REVIEW

File: driver/igemm_wrw_gtc_driver.h [hunk 1]
```diff
@@ -269,12 +270,12 @@ class igemm_wrw_gtc_t {
...
             return false;
 
-        if ((tunable->nxb != 1) && (tunable->nxe != 0)){
-            //std::cout << __func__ << " false: tunable->nxb is " << tunable->nxb << ", tunable->nxe is " << tunable->nxe << std::endl;
-            return false;
-        }
+        // if ((tunable->nxb != 1) && (tunable->nxe != 0)){
+        //     //std::cout << __func__ << " false: tunable->nxb is " << tunable->nxb << ", tunable->nxe is " << tunable->nxe << std::endl;
+        //     return false;
+        // }
```
Comment (shaojiewang) [inline driver/igemm_wrw_gtc_driver.h:276]: We may delete useless comments.

File: driver/igemm_wrw_gtc_driver.h [hunk 2]
```diff
@@ -254,7 +255,7 @@ class igemm_wrw_gtc_t {
 
         int gemm_m = k;
         int gemm_n = c * y * x;
-        int gemm_k = n * ho * wo;
+        int gemm_k = n * b;
 
         if ((gemm_n % gemm_n_per_block != 0) || (gemm_m % gemm_m_per_block !=0 )){
```
Comment (shaojiewang) [inline driver/igemm_wrw_gtc_driver.h:260]: I will pull an another request to change the condition.

File: igemm/algo/igemm_wrw_gtc.py [hunk 3]
```diff
@@ -87,16 +87,23 @@ def name(self):
...
-        self._emit(f"v_mul_lo_u32 v[{self.v_tmp(2)}], s[{self.s_dilation_h()}], v[{self.v_wei_iy()}]")
-        self._emit(f"v_mul_lo_u32 v[{self.v_tmp(3)}], s[{self.s_dilation_w()}], v[{self.v_wei_ix()}]")
-        self._emit(f"; transform hip, wip -> hi, wi")
-        self._emit(f"v_add_u32 v[{self.v_tmp()}], v[{self.v_tmp()}], v[{self.v_tmp(2)}]")
-        self._emit(f"v_add_u32 v[{self.v_tmp(1)}], v[{self.v_tmp(1)}], v[{self.v_tmp(3)}]")
-        self._emit(f"v_sub_i32 v[{self.v_in_ihi()}], v[{self.v_tmp()}], s[{self.s_pad_h()}]")
-        self._emit(f"v_sub_i32 v[{self.v_in_iwi()}], v[{self.v_tmp(1)}], s[{self.s_pad_w()}]")
+        self._emit(f"v_add_i32 v[{self.v_in_ihi()}], v[{self.v_tmp()}], v[{self.v_wei_iy()}]")
+        self._emit(f"v_mul_lo_u32 v[{self.v_tmp(1)}], s[{self.s_stride_w()}], v[{self.v_out_iwo()}]")   
+        self._emit(f"v_add_i32 v[{self.v_in_iwi()}], v[{self.v_tmp(1)}], v[{self.v_wei_ix()}]")
```
Comment (shaojiewang) [inline igemm/algo/igemm_wrw_gtc.py:95]: Can here use a single mad_i32_i24 instruction?

File: igemm/algo/igemm_wrw_gtc.py [hunk 4]
```diff
@@ -87,16 +87,23 @@ def name(self):
...
+        # self._emit(f"; transform iho, iwo, iy, ix -> hip, wip")
+        # self._emit(f"v_mul_lo_u32 v[{self.v_tmp()}], s[{self.s_stride_h()}], v[{self.v_out_iho()}]")
+        # self._emit(f"v_mul_lo_u32 v[{self.v_tmp(1)}], s[{self.s_stride_w()}], v[{self.v_out_iwo()}]")
+        # self._emit(f"v_mul_lo_u32 v[{self.v_tmp(2)}], s[{self.s_dilation_h()}], v[{self.v_wei_iy()}]")
+        # self._emit(f"v_mul_lo_u32 v[{self.v_tmp(3)}], s[{self.s_dilation_w()}], v[{self.v_wei_ix()}]")
+        # self._emit(f"; transform hip, wip -> hi, wi")
+        # self._emit(f"v_add_u32 v[{self.v_tmp()}], v[{self.v_tmp()}], v[{self.v_tmp(2)}]")
+        # self._emit(f"v_add_u32 v[{self.v_tmp(1)}], v[{self.v_tmp(1)}], v[{self.v_tmp(3)}]")
+        # self._emit(f"v_sub_i32 v[{self.v_in_ihi()}], v[{self.v_tmp()}], s[{self.s_pad_h()}]")
+        # self._emit(f"v_sub_i32 v[{self.v_in_iwi()}], v[{self.v_tmp(1)}], s[{self.s_pad_w()}]")
```
Comment (shaojiewang) [inline igemm/algo/igemm_wrw_gtc.py:106]: We may delete useless comments.

File: driver/igemm_wrw_gtc_driver.h (hunk 1 above)
Comment (carlushuang) [reply driver/igemm_wrw_gtc_driver.h:276]: OK

File: driver/igemm_wrw_gtc_driver.h (hunk 2 above)
Comment (carlushuang) [reply driver/igemm_wrw_gtc_driver.h:260]: OK

File: igemm/algo/igemm_wrw_gtc.py (hunk 3 above)
Comment (carlushuang) [reply igemm/algo/igemm_wrw_gtc.py:95]: For this PR I think just keep the mul_lo_u32, to make this safer. In future re-factor we may consider change to i32_i24

File: igemm/algo/igemm_wrw_gtc.py (hunk 4 above)
Comment (carlushuang) [reply igemm/algo/igemm_wrw_gtc.py:106]: OK

File: igemm/algo/igemm_wrw_gtc.py (hunk 3 above)
Comment (shaojiewang) [reply igemm/algo/igemm_wrw_gtc.py:95]: Agreed

**Review by shaojiewang — APPROVED** [review]

**Pull request merged**